lib/Entity/Tag.php

```diff
@@ -542,4 +735,101 @@ private function unlinkDisplayGroups()
 
         $this->getStore()->update($sql, $params);
     }
+
+    /**
+     * Is this tag a system tag?
+     * @return bool
+     * @throws \Xibo\Exception\NotFoundException
+     */
+    public function isSystemTag()
+    {
+        $tag = $this->tagFactory->getById($this->tagId);
+
+        return $tag->isSystem === 1;
+    }
+
+    /**
+     * @throws InvalidArgumentException
+     */
+    public function validateTagOptions()
+    {
+        if (isset($this->options)) {
+            $tagOptionsString = implode(',', json_decode($this->options));
+            $tagOptionsArray = explode(',', $tagOptionsString);
+
+            if (isset($this->value) && !in_array($this->value, $tagOptionsArray)) {
+                throw new InvalidArgumentException(sprintf(__('Provided tag value %s, not found in tag %s options',
+                    $this->value, $this->tag)), 'tags');
+            }
+
+            if (!isset($this->value) && $this->isRequired == 1) {
+                throw new InvalidArgumentException(sprintf(__('Provided Tag %s , requires a value', $this->tag)),
+                    'tags');
+            }
+        }
+
+        if(isset($this->value) && !isset($this->options)) {
+            throw new InvalidArgumentException(sprintf(__('Provided Tag %s, does not have defined option values', $this->tag)), 'tags');
+        }
+
+    }
+
+    /**
+     * Removes Tag value from lktagtables
+     *
+     * @param array $values An Array of values that should be removed from assignment
+     */
+    public function updateTagValues($values)
+    {
+        $this->getLog()->debug('Tag options were changed, the following values need to be removed ' . json_encode($values));
+
+        foreach ($values as $value) {
+            $this->getLog()->debug('removing following value from lktag tables ' . $value);
+
+            if ($this->getStore()->exists('SELECT * FROM `lktagcampaign` INNER JOIN `tag` on lktagcampaign.tagId = tag.tagId WHERE `tag` = :tag AND value = :value', ['value' => $value, 'tag' => $this->tag])) {
+
+                $this->getStore()->update('UPDATE `lktagcampaign` SET `value` = null WHERE tagId = :tagId AND value = :value',
```

Comment:
I dont think you need to do any of the prior checking - you can just run the update statement because if the where clause doesn't match, it won't update anything.